dataset/src/main/scala/frameless/functions/NonAggregateFunctions.scala

```diff
@@ -70,40 +64,36 @@ trait NonAggregateFunctions {
     *   [[https://github.com/apache/spark/blob/4a3c09601ba69f7d49d1946bb6f20f5cfe453031/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala#L67]]
     * apache/spark
     */
-  def atan2[A, B, T](l: TypedColumn[T,A], r: TypedColumn[T, B])
+  def atan2[A, B, T](l: AbstractTypedColumn[T,A], r: AbstractTypedColumn[T, B])
     (implicit
       evCanBeDoubleL: CatalystCast[A, Double],
       evCanBeDoubleR: CatalystCast[B, Double]
-    ): TypedColumn[T, Double] = {
-      implicit val lUnencoder = l.uencoder
-      implicit val rUnencoder = r.uencoder
-      new TypedColumn[T, Double](untyped.atan2(l.cast[Double].untyped, r.cast[Double].untyped))
-    }
+    ): r.ThisType[T, Double] =
+    r.typed(untyped.atan2(l.cast[Double].untyped, r.cast[Double].untyped))
 
-  def atan2[B, T](l: Double, r: TypedColumn[T, B])(implicit evCanBeDoubleR: CatalystCast[B, Double]): TypedColumn[T, Double] =
-    atan2(lit(l): TypedColumn[T, Double], r)
+  def atan2[B, T](l: Double, r: AbstractTypedColumn[T, B])
+     (implicit
+      evCanBeDoubleR: CatalystCast[B, Double]): r.ThisType[T, Double] = atan2(r.lit(l), r)
 
-  def atan2[A, T](l: TypedColumn[T, A], r: Double)(implicit evCanBeDoubleL: CatalystCast[A, Double]): TypedColumn[T, Double] =
-    atan2(l, lit(r): TypedColumn[T, Double])
+  def atan2[A, T](l: AbstractTypedColumn[T, A], r: Double)
+     (implicit
+      evCanBeDoubleL: CatalystCast[A, Double]): l.ThisType[T, Double] =
+    atan2(l, l.lit(r)).asInstanceOf[l.ThisType[T, Double]]
```

Comment:
Why is this cast neeeded?

Reply:
Good question, idk, but it doesn't compile otherwise. Probably some type inference quark. It works for the other one that has the right parameter to be `AbstractTypedColumn[T, A]` but it doesn't work for this one her. I am kind of "helping" the compiler with the asInstanceOf. If you have this branch already locally, can you give it a shot?

Reply:
It's because it calls into `atan2` which doesn't check both l & r have same the same type. I'll fix it